### DISCUSSION TITLE
[Book1][Section 9.4] True Lambertian Reflection not following the Lambertian cos law ?

### DISCUSSION THREAD

**Discussion by Imasu** [RayTracing/raytracing.github.io#1380] [Q&A] [answered]
Hi,

The book is great and it allows me to both refresh my maths & and my C#. 

I can’t however understand mathematically the explanation given to obtain a diffuse scattering following the Lambert cosine law. This law states that the probability of reflection follows a cosine law centered on the normal: there is more probability that the reflected ray is close to the normal (if angle theta between the normal and this vector ~ 0 -> probability ~1) than close to the tangent plane (angle ~90 -> probability ~0). It is indeed a cosine.

The implementation proposed advocates the use of a random unit vector added to the normal (N) to define the direction of the reflected ray (R). I don’t see how statistically a cosine law would apply: in my opinion, the probability of drawing the unit ray is the same in all directions, and therefore follows a linear law. If we restrict the analysis to a 2D space, it is easy to model it by calculating the angle of each resulting ray with the normal: they have the same probabilities of occurence.

I found on the internet this page https://www.particleincell.com/2015/cosine-distribution/  that explains how to obtain a cosine law. By remodeling it in Excel and using a online 3D viewer, it works correctly. The probability of drawing is higher around the normal.

The implementation of the code is relatively simple. What is troubling, however, is that the resulting image is very similar to the one obtained by the book method. I wonder if differences would only appear with a more complex world.

Can you tell me where I went wrong in my reasoning?
Thank you.

Lambertian simulation:  (3D online viewer : https://miabellaai.net)
![Capture d'écran 2024-02-08 215206](https://github.com/RayTracing/raytracing.github.io/assets/4583573/16356c70-90fd-41f5-961d-4dfc3196192f)

C# code for the method:
        public static Vec3 RandomUnitVec3LambertDistrib(Vec3 normal) {

            // Rotation vector
            double thetaSin = Math.Sqrt(MathUtil.RandomDouble());
            double thetaCos = Math.Sqrt(1 - thetaSin * thetaSin);
            double psi = MathUtil.RandomDouble() * 2 * Math.PI;
            double psiSin = Math.Sin(psi);
            double psiCos = Math.Cos(psi);
            Vec3 vectRotation = new(thetaSin * psiCos, thetaSin * psiSin, thetaCos);

            // Unitary vectors on the surface
            //    https://stackoverflow.com/questions/33658620/generating-two-orthogonal-vectors-that-are-orthogonal-to-a-particular-direction
            //    https://en.wikipedia.org/wiki/Cross_product#:~:text=In%20mathematics%2C%20the%20cross%20product%20or%20vector%20product,three-dimensional%20space%2C%20and%20is%20denoted%20by%20the%20symbol
            Vec3 vectNormalUnit = UnitVec3(normal);
            Vec3 vectRandomUnit = RandomUnitVec3();
            Vec3 vectOrtho1 = UnitVec3(vectRandomUnit - Vec3.Dot(vectRandomUnit, vectNormalUnit) * vectNormalUnit);
            Vec3 vectOrtho2 = Vec3.Cross(vectNormalUnit, vectOrtho1);

            return vectRotation.X * vectOrtho1 + vectRotation.Y * vectOrtho2 + vectRotation.Z * vectNormalUnit;
        }

(I used a french / english translator software )

**Comment by armansito** [comment, marked as answer]
I'll try to give an intuitive answer to your question without going too deep into the math (at least based on my understanding of the topic). It's true that the samples are spread uniformly on the surface of the unit sphere but this doesn't directly represent the distribution of ray directions and offsetting the vectors by the normal vector introduces a bias to the overall distribution of scattering directions. I think it's more intuitive to think about the integral over all scattering directions as a hemisphere centered at the point of intersection and aligned to the normal vector, such that all points on the surface of the hemisphere are equally distant from the intersection point:

<img width="1114" alt="Screenshot 2024-02-09 at 6 31 01 PM" src="https://github.com/RayTracing/raytracing.github.io/assets/6933700/3cf49933-4ce5-4bee-afac-165957ffd31f">

Here the orange circle represents the sampled (and offset) unit sphere and the black semi-circle represents the hemisphere of ray directions. If you project the uniformly distributed set of points on the sphere (orange in the figure above) onto the hemisphere, you'll find that their overall distribution looks a lot like the picture you posted above. I put together a Python script and created a plot to demonstrate this:

Sphere Samples             |  Ray Directions, projected
:-------------------------:|:-------------------------:
![](https://github.com/RayTracing/raytracing.github.io/assets/6933700/cc7bef69-5c90-47e0-87e6-364e30b50403)  |  ![](https://github.com/RayTracing/raytracing.github.io/assets/6933700/790d0be8-1dc6-43ae-b44a-d7b7f3e569ce)

The point of intersection is located at $(0, 0, -1)$ and every blue point corresponds to an orange point that was projected along a unit vector relative to $(0, 0, -1)$.

The blue dome corresponds to a biased (cosine-weighted) BRDF. I tried to provide a visual intuition but there is quite a bit of theory behind why the hemisphere makes sense for the reflectance model and how the cosine-weighted bias relates to the cosine term in the rendering equation. I recommend the following as further reading on the general theory behind this:

- [Ray Tracing: The Rest of Your Life, Chapter 4: Monte-Carlo Integration on the Sphere of Directions](https://raytracing.github.io/books/RayTracingTheRestOfYourLife.html#montecarlointegrationonthesphereofdirections)
- [Crash Course in BRDF Implementation](https://boksajak.github.io/files/CrashCourseBRDF.pdf) by Jakub Boksansky
- The chapters on [Reflection Models](https://www.pbr-book.org/4ed/Reflection_Models) and [Sampling Multidimensional Functions](https://www.pbr-book.org/4ed/Sampling_Algorithms/Sampling_Multidimensional_Functions) sections in the PBRT book (the latter presents an alternative method for cosine-weighted hemisphere sampling which you may also find interesting).
- The [Materials chapter](https://graphicscodex.courses.nvidia.com/app.html?page=_rn_matrls) in the Graphics Codex.

I hope this helps!

P.S.:
---------
Higher-density plots of the projected points (showing both hemispheres for clarity):

<img width="815" alt="Screenshot 2024-02-09 at 10 24 30 PM" src="https://github.com/RayTracing/raytracing.github.io/assets/6933700/0f9915a6-5ff0-4a2f-a7ae-d2037e81af32">

---------
Looking down the normal vector:

<img width="1041" alt="Screenshot 2024-02-09 at 10 25 45 PM" src="https://github.com/RayTracing/raytracing.github.io/assets/6933700/c6372b00-cee3-4cf8-aaa8-98c76d0743f9">

**Reply by Imasu** [reply]
Thank you so much for taking time to answer to me.
I was able to simulate the projection of the direction vectors and, indeed, the result is similar to the one I provided.

**Reply by dimitry-ishenko** [reply]
@armansito thanks from me as well. I was about to ask similar question.
